### PR TITLE
refactor(prover): refactor prover / guardian prover

### PR DESCRIPTION
**Pull request by davidtaikocha**
`BlockProposed` event handler main workflow:
image

`TransitionProved` event handler main workflow:
image

`TransitionContested` event handler main workflow:
image

`AssignmentExpired` event handler main workflow:
image

### CODE REVIEW

File: prover/event_handler/util.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	if err != nil {
+		return false, 0, fmt.Errorf("failed to get proving window: %w", err)
+	}
+
+	var (
+		now       = uint64(time.Now().Unix())
+		exipredAt = e.Meta.Timestamp + uint64(provingWindow.Seconds())
+	)
+
+	return now > exipredAt, time.Duration(exipredAt-now) * time.Second, nil
```
Comment (cyberhorsey) [inline prover/event_handler/util.go:155]: typo, `exipredAt`, should be `expiredAt`

File: prover/event_handler/util.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	proposedIn *big.Int,
+) (*bindings.TaikoDataBlockMetadata, error) {
+	e, err := getBlockProposedEventFromBlockID(ctx, rpc, id, proposedIn)
+	if err != nil {
+		return nil, err
+	}
+	return &e.Meta, nil
+}
+
+// isProvingWindowExpired returns true if the assigned prover proving window of
```
Comment (cyberhorsey) [inline prover/event_handler/util.go:139]: maybe update comment to say the second return paramter is the time remaining til proving window is expired

**Review by cyberhorsey — CHANGES_REQUESTED** [review]

**Review by YoGhurt111 — APPROVED** [review]

File: prover/server/api.go [hunk 3]
```diff
@@ -105,25 +105,23 @@ func (srv *ProverServer) CreateAssignment(c echo.Context) error {
...
-			log.Warn(
-				"Insufficient prover balance, please get more tokens or wait for verification of the blocks you proved",
-				"prover", srv.proverAddress,
-			)
-			return echo.NewHTTPError(http.StatusUnprocessableEntity, "insufficient prover balance")
-		}
+	if !ok {
+		log.Warn(
+			"Insufficient prover balance, please get more tokens or wait for verification of the blocks you proved",
+			"prover", srv.proverAddress,
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply prover/server/api.go:122]: Emm looks like there are logs in the above `CheckProverBalance`, so its ok?

File: prover/init.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+		}
+
+		startingBlockID = new(big.Int).SetUint64(stateVars.B.LastVerifiedBlockId)
+	}
+
+	log.Info("Init L1Current cursor", "startingBlockID", startingBlockID)
+
+	latestVerifiedHeaderL1Origin, err := p.rpc.L2.L1OriginByID(p.ctx, startingBlockID)
+	if err != nil {
+		if err.Error() == ethereum.NotFound.Error() {
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply prover/init.go:184]: Changed

File: prover/event_handler/assignment_expired.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/taikoxyz/taiko-client/bindings"
+	"github.com/taikoxyz/taiko-client/pkg/rpc"
+	proofSubmitter "github.com/taikoxyz/taiko-client/prover/proof_submitter"
+)
+
+// AssignmentExpiredEventHandler is responsible for handling the expiration of proof assignments.
+type AssignmentExpiredEventHandler struct {
+	rpc               *rpc.Client
+	proverAddress     common.Address
+	proofSubmissionCh chan *proofSubmitter.ProofRequestBody
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply prover/event_handler/assignment_expired.go:18]: Changed for all event handlers

File: prover/event_handler/assignment_expired.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	proofContestCh    chan *proofSubmitter.ContestRequestBody
+	contesterMode     bool
+}
+
+// NewAssignmentExpiredEventHandler creates a new AssignmentExpiredEventHandler instance.
+func NewAssignmentExpiredEventHandler(
+	rpc *rpc.Client,
+	proverAddress common.Address,
+	proofSubmissionCh chan *proofSubmitter.ProofRequestBody,
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply prover/event_handler/assignment_expired.go:27]: Changed for all event handlers

File: prover/event_handler/transition_proved.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+) error {
+	metrics.ProverReceivedProvenBlockGauge.Update(e.BlockId.Int64())
+
+	// If this prover is in contest mode, we check the validity of this proof and if it's invalid,
+	// contest it with a higher tier proof.
+	if !h.contesterMode {
+		return nil
+	}
+
+	isValidProof, err := isValidProof(
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply prover/event_handler/transition_proved.go:45]: Changed to `isValid`

File: prover/event_handler/transition_proved.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+		return err
+	}
+
+	meta, err := getMetadataFromBlockID(ctx, h.rpc, e.BlockId, new(big.Int).SetUint64(blockInfo.Blk.ProposedIn))
+	if err != nil {
+		return err
+	}
+
+	log.Info(
+		"Contest a proven transition",
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply prover/event_handler/transition_proved.go:73]: Changed

File: prover/event_handler/util.go (hunk 1 above)
Comment (davidtaikocha) [reply prover/event_handler/util.go:155]: Changed

File: prover/event_handler/util.go (hunk 2 above)
Comment (davidtaikocha) [reply prover/event_handler/util.go:139]: changed

File: prover/prover.go [hunk 9]
```diff
@@ -478,346 +375,68 @@ func (p *Prover) proveOp() error {
...
-func (p *Prover) onBlockProposed(
-	ctx context.Context,
-	event *bindings.TaikoL1ClientBlockProposed,
-	end eventIterator.EndBlockProposedEventIterFunc,
-) error {
-	// If we are operating as a guardian prover,
-	// we should sign all seen proposed blocks as soon as possible.
+// contestProofOp performs a proof contest operation.
+func (p *Prover) contestProofOp(ctx context.Context, req *proofSubmitter.ContestRequestBody) {
 	go func() {
```
Comment (mask-pp) [inline prover/prover.go:380]:
```suggestion
	p.wg.Add(1)
	go func() {
	defer p.wg.Done()
```

**Review by mask-pp — APPROVED** [review]

**Review by cyberhorsey — APPROVED** [review]

**Pull request merged**